docs/design/configs.md

```diff
@@ -0,0 +1,130 @@
+# Generic Config Object
```

Comment:
We need to think a little about how this would interact with template parameter substitution, as discussed in #4210 and #1743.